report/verification/mms/simple_transient.md

```diff
@@ -0,0 +1,312 @@
+    format_name: myst
+    format_version: 0.13
+    jupytext_version: 1.16.2
+kernelspec:
+  display_name: vv-festim-report-env
+  language: python
+  name: python3
+---
+
+# Simple transient diffusion case
+
+```{tags} 2D, MMS, transient
+```
+
+This is a simple transient MMS example.
+We will only consider diffusion of hydrogen in a unit square domain $\Omega$ at steady state with an homogeneous diffusion coefficient $D$.
+Moreover, a Dirichlet boundary condition will be assumed on the boundaries $\partial \Omega $.
+
+The problem is therefore:
+
+$$
+\begin{align}
+    &\nabla \cdot (D \ \nabla{c}) - \frac{\partial c}{\partial t} = -S  \quad \text{on }  \Omega  \\
+    & c = c_0 \quad \text{on }  \partial \Omega
+\end{align}
+$$(problem_simple_transient)
+
+The exact solution for mobile concentration is:
+
+$$
+\begin{equation}
+    c_\mathrm{exact} = 1 + 2 x^2 + 3 y^2 t + 2t
+\end{equation}
+$$(c_exact_simple_transient)
+
+Injecting {eq}`c_exact_simple_transient` in {eq}`problem_simple_transient`, we obtain the expressions of $S$ and $c_0$:
+
+\begin{align}
+    & S = 2 + 3 y^2 - (4 + 6t) D \\
+    & c_0 = c_\mathrm{exact}
+\end{align}
+
+We can then run a FESTIM model with these values and compare the numerical solution with $c_\mathrm{exact}$.
+
++++
+
+## FESTIM code
+
+```{code-cell} ipython3
+:tags: [hide-cell]
+import festim as F
+import sympy as sp
+import fenics as f
+import matplotlib as mpl
+import matplotlib.pyplot as plt
+import numpy as np
+
+# Create and mark the mesh
+nx = ny = 100
+fenics_mesh = f.UnitSquareMesh(nx, ny)
+
+
+volume_markers = f.MeshFunction("size_t", fenics_mesh, fenics_mesh.topology().dim())
+volume_markers.set_all(1)
+
+surface_markers = f.MeshFunction(
+    "size_t", fenics_mesh, fenics_mesh.topology().dim() - 1
+)
+surface_markers.set_all(0)
+
+class Boundary(f.SubDomain):
+    def inside(self, x, on_boundary):
+        return on_boundary
+
+boundary = Boundary()
+boundary.mark(surface_markers, 1)
+
+# Create the FESTIM model
+my_model = F.Simulation()
+
+my_model.mesh = F.Mesh(
+    fenics_mesh, volume_markers=volume_markers, surface_markers=surface_markers
+)
+
+# Variational formulation
+exact_solution = (
+    1 + 2 * F.x**2 + 3 * F.t * F.y**2 + 2 * F.t
+)  # exact solution
+
+exact_solution_copy = exact_solution.subs(F.x, F.x)
+
+D = 2
+
+my_model.sources = [
+    F.Source(2 + 3 * F.y**2 - (4 + 6 * F.t) * D, volume=1, field="solute"),
+]
+
+my_model.boundary_conditions = [
+    F.DirichletBC(surfaces=[1], value=exact_solution, field="solute"),
+]
+
+my_model.materials = F.Material(id=1, D_0=D, E_D=0)
+
+my_model.T = F.Temperature(500)  # ignored in this problem
+
+xdmf_file_name = "simple_transient_mobile.xdmf"
+my_model.exports = [F.XDMFExport(field="solute", filename=xdmf_file_name, checkpoint=True)]
+
+final_time = 16
+slices = 4 # <-------- free parameter, better if slices | final_time
+slice_size = final_time // slices
+milestones = [slice_size * i for i in range(1, slices + 1)]
```

Comment:
I would simplify this to something like:

```suggestion
milestones = np.linspace(0, final_time, num=4)[1:]
```

Reply:
note: passing a numpy array to `F.Stepsize` crashes FESTIM due to [this line](https://github.com/festim-dev/FESTIM/blob/a2fd832f83b952c812b2e8abc2c63d8297ce9261/festim/stepsize.py#L75), where the error is `ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()`.

I think this would be a good place to do type checking in the future but for now, if you asked me how to fix it, I'd do
```python
@milestones.setter
def milestones(self, value = None):
    if len(value):
        self._milestones = sorted(value)
    else:
        self._milestones = value
```

As for the change, I'm going to implement it and try something to circumvent the precision issue, which isn't that hard to fix in retrospect.

Reply:
It was not hard to fix.

Reply:
I think it would be better to have:

```python
if value is not None:
```